### PR TITLE
Next higher number with same set bits implemented

### PR DESCRIPTION
**Pull request by Kunal766**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
+/**
+ * @file
+ * @brief [Next higher number with same number of set bits]
+ * (https://www.geeksforgeeks.org/next-higher-number-with-same-number-of-set-bits/)
+ * implementation
+ *
+ * @details
+ * Given a number x, find next number with same number of 1 bits in it’s binary representation.
+ * For example, consider x = 12, whose binary representation is 1100 (excluding leading zeros on 32 bit machine).
+ It contains two logic 1 bits. The next higher number with two logic 1 bits is 17 (100012).
```
Comment (Panquesito7) [inline bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp:10]:
```suggestion
 * It contains two logic 1 bits. The next higher number with two logic 1 bits is 17 (100012).
```

File: bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#include <iostream>  /// for IO operations
+
+/**
+ * @namespace bit_manipulation
+ * @brief Bit manipulation algorithms
+ */
+namespace bit_manipulation {
+/**
+ * @brief The main function implements check next numbet
+ * @param x is the number who's next number will be calculated
```
Comment (Panquesito7) [inline bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp:27]:
```suggestion
 * @brief The main function implements checking the next number
 * @param x the number that will be calculated
```

File: bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ * @returns a number
+ */
+
+int next_higher_number(int x)
```
Comment (Panquesito7) [inline bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp:31]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help here or on our [Discord server](https://the-algorithms.com/discord).** 🙂

File: bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    assert(bit_manipulation::next_higher_number(64) == 128);
+    // x = 15 return 23
+    assert(bit_manipulation::next_higher_number(15) == 23);
+    // x= 32 return 64
+    assert(bit_manipulation::next_higher_number(32) == 64);
+    // x = 97 return 98
+    assert(bit_manipulation::next_higher_number(97) == 98);
+    // x = 1024 return 2048
+    assert(bit_manipulation::next_higher_number(1024) == 2048);
+    std::cout << "All test cases successfully passed!" << std::endl;
```
Comment (Panquesito7) [inline bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp:93]:
```suggestion

    std::cout << "All test cases have successfully passed!" << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]
Amazing. Thanks! 🚀

**Review by realstealthninja — CHANGES_REQUESTED** [review]
bit of it a nit pick, but thank you so much for your contribution!

File: bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * @brief The main function implements checking the next number
+ * @param x the number that will be calculated
+ * @returns a number
+ */
+
+uint64_t next_higher_number(uint64_t x)
+{
+ 
+  uint64_t rightOne;
```
[the remark this answers is not in the document]
Comment (Panquesito7) [reply bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp:34]: We should probably use `snake_case`. I'll create a discussion in the repository. 👍

**Review by realstealthninja — APPROVED** [review]
LGTM

File: bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Panquesito7) [reply bit_manipulation/next_higher_number_with_same_number_of_set_bits.cpp:34]: What should we do about this? Should we keep it as-is or?

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**